### ISSUE TITLE
Definition of Backwards Compatibility

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-thing-description#1243]
We talk a lot about TD 1.1 having the "backwards compatibility" feature. However, the only place that sort of mentions this is in the abstract with the following sentence:

> This specification describes a superset of the features defined in Thing Description 1.0 [WOT-THING-DESCRIPTION]. Unless otherwise specified, documents created with version 1.0 of this specification remain compatible with Thing Description 1.1. 

To me, this is not clear. What does it mean that **a** TD is compatible with another one? I think this should be defined somewhere in the introduction with a clear sentence that is not lost in a sentence. From what we talk in the meetings, I think what most people think it means is something like:

Documents that conform to TD 1.1 should not cause validation issues in TD 1.0 consumers and TD 1.1 aware consumers should still accept TD 1.0 documents.

What it cannot mean is :
A TD 1.0 consumer can totally understand a TD 1.1 document. (We are adding new keywords that a TD 1.0 consumer does not know about)

What is somewhat ambiguous to me still are:
- Cases where we add a new keyword that can replace an old one but we decide to keep the old one until TD 2.0. For example, if we say that you can use now use `schema` in properties to encapsulate the DataSchema (like output of actions), TD 1.1 implementors may choose to do this and put the data schema of a property there and not inside the property. For a TD 1.0 consumer, it will look as if the property does not return any data or an unstructured data.
- When we update `@context`, it will break validators of TD 1.0. Does this mean that we should have now an `@context` that looks like this: `"@context": ["https://www.w3.org/2019/wot/td/v1","https://www.w3.org/2022/wot/td/v11"],` ? This is a valid context for TD1.0 implementors. Similarly, new `op` values will produce validation errors.

**Comment by danielpeintner** [comment]
My point was different: What I was meaning was that IF I know that I have to deal with 1.0 consumers in production (or I simply want to be on the safe side) I can simply convert any TD in the according 1.0 version *always*.

It would be good to have a mean to identify the consumer (and its support).

**Comment by egekorkan** [comment]
I would say that the conversion is another topic. We should fix this definition and then worry about conversion later on since there will be producers and consumers of v1.0 or of v1.1 who cannot convert one to the other one.

After thinking for a couple of days, I think that the second part of my original comment is the most realistic one, i.e.:

> TD 1.1 aware consumers should still accept TD 1.0 documents.

E.g. if TD1.1 defines `subscribeallevents` but I am met with at TD1.0 document that has 100 events that need separate subscription, I won't have this convenience operation. 

This means that we can have:
- schema keyword in affordances that is optional
- new op values
- new action behavior via hypermedia etc.
- `rel` values
- completely new `@context` -> TD1.1 schema must have both contexts with an oneOf in the schema
- we can have inline security definitions -> TD1.1 consumers must support both ways

We cannot have:
- inline security defitions by removing the old way
- less op values

In other words, if we introduce a new way to do something, the old way should still be kept. This also implies that if someone wants to write a generic TD1.1 consumer, they will be trying to compliant with two spec versions which makes it annoying. Probably, they should wait for TD2.0

**Comment by sebastiankb** [comment]
from today's TD call:
- the group like to stay to be backwards compatible to 1.0
- when we add new terms / structures that will not have any impact on any mandatory terms/structures that are defined in the 1.0 (e.g., there will be no redefinition of array type based terms to objects)
- an exception will be only the new context file that comes with a new namespace like `https://www.w3.org/2022/wot/td/11/v1`

**Comment by benfrancis** [comment]
> * the group like to stay to be backwards compatible to 1.0

I'm still not clear on what the definition of "backwards compatible" is here. Is it "1.0 Consumer interacting with a 1.1 Thing" or "1.1 Consumer interacting with a 1.0 Thing"?

Which of the changes I listed above would be lost if you "Roll back TD spec to 1.1 (1.0 compatible) features" as proposed in the [WG 2021 extension plan](https://github.com/w3c/wot/pull/1000)?

**Comment by egekorkan** [comment]
I think that my comment did not end up here but I had written it in IRC (https://www.w3.org/2021/11/03-wot-td-minutes.html)

> maybe we can say that we add new terms, do not change the structure of the old terms but add new values. This means that if a TD parser was iterating through a key's values, it will not fail but it will meet new values.
… so for example, op will be an array and can be iterated through like an array but will have new values like `subscribeallevents`
… maybe @context needs some special note

**Comment by egekorkan** [comment]
This will mean that a 1.0 Consumer will be able to interact with 1.1 Thing if no validation is done (since it is not mandatory). Existing data structures should be parseable in the old way. 1.1 Consumer will be able to interact with a 1.0 Thing as well.

In both cases, `@context` can be of an issue so it needs a special note maybe

**Comment by benfrancis** [comment]
OK, so if I've understood correctly that allows:
* New members (e.g. `profile`, `additionalResponses`, `sizes`, `model`, `default`, `exclusiveMinimum`, `exclusiveMaximum`, `pattern`) and possible values of members (e.g. `queryaction`, `cancelaction`, `observeallproperties`, `unobserveallproperties`, `subscribeallevents`, `unsubscribeallevents`, `queryallactions`, `uri`)
* New value of `@context` for a 1.1 TD (not landed yet, but assumed)

so what about:
* Changed or added defaults (e.g. of `op`, `success`)
* Changed types (e.g. of `contentType`, `contentCoding`, `security`, `op`)
* Change in mandatory status of members (e.g. `forms`, not landed yet but under discussion) and disallowed members (e.g. `htv:methodName`)
* New namespace for Thing Models (`http://www.w3.org/ns/tm`)
* New canonicalisation feature

**Comment by egekorkan** [comment]
Commenting directly:
* Changed or added defaults (e.g. of `op`, `success`) : success is fine, just a new term that a 1.0 parser would ignore, not sure about op, it was already there. A new op value would be also ignored since a 1.0 consumer was not searching for it anyway.
* Changed types (e.g. of `contentType`, `contentCoding`, `security`, `op`) : These have not changed
* Change in mandatory status of members (e.g. `forms`, not landed yet but under discussion) and disallowed members (e.g. `htv:methodName`) : This would not be possible
* New namespace for Thing Models (`http://www.w3.org/ns/tm`) : No problem since a 1.0 consumer is not aware of TMs
* New canonicalisation feature : not sure how it effects a 1.0 consumer, needs investigation

**Comment by benfrancis** [comment]
I've realised that the list of changes I was referring to was [from the FPWD to the current draft](https://w3c.github.io/wot-thing-description/#changes-from-fpwd) rather than changes from 1.0 to 1.1, so that changes things slightly (sorry for the confusion).

> * 1 Changed or added defaults (e.g. of `op`, `success`) : success is fine, just a new term that a 1.0 parser would ignore, not sure about op, it was already there. A new op value would be also ignored since a 1.0 consumer was not searching for it anyway.

It was the default value of `success` that was changed since FPWD, but `AdditionalExpectedResponse` didn't exist at all in 1.0 anyway so it's fine. Similarly with `op`, it was the default value that has changed for `PropertyAffordance` forms, so that the default depends on the values of the `readOnly` and `writeOnly` members. This could technically result in a 1.0 consumer and 1.1 consumer interpreting the same Thing Description differently, but this was arguably a bug in 1.0 anyway, since a the Form of a `readOnly` property would have a default `op` member which includes `writeproperty`. A Producer which didn't override this default would have caused problems regardless.

>     * 2 Changed types (e.g. of `contentType`, `contentCoding`, `security`, `op`) : These have not changed

I see now that they were changed in the FPWD, but have since been changed back in the latest draft.

In addition to this there are [changes between the 2020 recommendation and the FPWD](https://www.w3.org/TR/2020/WD-wot-thing-description11-20201124/#changes-from-recommendation):

- List of content types that MAY be used with a data schema (this is fine as they're just suggestions, not an exclusive list)
- New terms `contentEncoding`, `contentMediaType`, `multipleOf`, `minLength`, `maxLength`, a new security scheme called `ComboSecurityScheme` and new flows for the `OAuth2SecurityScheme` (presumably these would all be ignored by a 1.0 consumer and just not used by a 1.0 producer)
- Default value definition given for op term in `EventAffordance` forms changed from `subscribeevent` to `['subscribeevent', 'unsubscribeevent']` (technically this could result in a 1.0 Consumer and 1.1 Consumer interpreting the same Thing Description differently so that a 1.1 Consumer assumes that the event has an unsubscribe operation. This may not make much difference in practice since if unsubscribe is an explicit operation it probably requires additional metadata in order to be used anyway).

In conclusion, assuming that a 1.0 Consumer will ignore any new members and values in a 1.1 Thing Description (which probably still needs testing), the biggest problems I can see are:
1. Strictly speaking a 1.0 Consumer will see a Thing Description as invalid if the `@context` is not equal to or does not start with `https://www.w3.org/2019/wot/td/v1`. Given the 1.0 specification is already fixed and implemented, this means that all 1.1 Thing Descriptions would have to include the 1.0 context first in order to be accepted by a 1.0 Consumer. I've proposed a solution to this [above](https://github.com/w3c/wot-thing-description/issues/1243#issuecomment-949600882).
2. It means that `Form`s remain mandatory at the interaction affordance level, so for example if a Web Thing uses a single WebSocket endpoint for all operations then it will have to duplicate that endpoint in Forms for every interaction affordance. There's still potentially a workaround for this if for example a Web Thing provides both HTTP and WebSocket endpoints (as is the case with most WoT Producers in the [WebThings](https://webthings.io/) platform), since the HTTP endpoint can satisfy the requirement and a single WebSocket endpoint can be provided in a top level form. As long as this bug/feature remains, then it isn't such a barrier to us making WebThings W3C compliant.

As a sidenote, hopefully the final 1.1 specification will include a list of changes compared with 1.0, since currently figuring out the net difference is a bit complicated!

**Comment by relu91** [comment]
> Strictly speaking a 1.0 Consumer will see a Thing Description as invalid if the @context is not equal to or does not start with https://www.w3.org/2019/wot/td/v1. Given the 1.0 specification is already fixed and implemented, this means that all 1.1 Thing Descriptions would have to include the 1.0 context first in order to be accepted by a 1.0 Consumer. I've proposed a solution to this above.

Is it really a problem? If we assumed that validation was not mandatory in TD 1.0 why a 1.0 would complain about the context string?

**Comment by egekorkan** [comment]
Exactly, validation of a TD is not mandatory for consumption. The JSON Schema at https://w3c.github.io/wot-thing-description/#json-schema-for-validation is informative. There are also no assertions at https://w3c.github.io/wot-thing-description/#validation-serialization-json which is even a new chapter that did not exist before

**Comment by benfrancis** [comment]
Whilst the specification doesn't say that a Thing Description has to be validated before being parsed, it does say that the `@context` member is mandatory and that:

> The `@context` name-value pair MUST contain the anyURI https://www.w3.org/2019/wot/td/v1 either directly when of type anyURI or as first element when of type Array.

It would therefore seem reasonable that a Consumer might reject a Thing Description if this assertion is not true, or depend on this value being set in order to detect that a given resource is a Thing Description in the first place.

For example, @relu91 [code you contributed to WebThings Gateway](https://github.com/WebThingsIO/gateway/pull/2880/files#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351) (which I reviewed) looks for this `@context` value in a Thing Description, and if not present will not try to parse it as a W3C WoT Thing Description.

As an aside, it has been troubling me for some time that there's no algorithm for parsing a Thing Description defined in the Thing Description specification. Maybe this is something that could be added in 2.0 to address these kinds of ambiguities.

**Comment by relu91** [comment]
You're right I keep forgetting that assertion, sorry. It really seems that we have to use your solution above to work around this issue. I'm wondering if it makes sense to have that as a RULE or something optional.... but I guess making it optional is a bit cumbersome.

**Comment by egekorkan** [comment]
Call of 15.06:
- The JSON Schema now accepts old TDs and the discussion is settled. Closing
